### ISSUE TITLE
"r10k puppetfile install" fails to update some modules.

### ISSUE THREAD

**Issue by TheKangaroo** [puppetlabs/r10k#767]
We use a setup where a Jenkins installs the modules from Puppetfile in the workspace with 
```
r10k puppetfile install -v
```
and then rsync's them to various puppetmasters.
I recently set up a new Jenkins with r10k v2.5.5 (old one was 2.1.1)
Now I see the strange behavior that r10k sometimes fails to update a single module in the catalog. 
```
INFO	 -> Updating module /var/lib/jenkins/workspace/puppet_code_deploy/{env}/modules/a_module
WARN	 -> Skipping /var/lib/jenkins/workspace/puppet_code_deploy/{env}/modules/a_module due to local modifications
```
I didn't do any modifications on the local repo, but saw this new git remote on the repo:
```
cache	/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git (push)
```
Since this is the same in every environment I thought maybe there are interferences with other envs.

I could remove the modules folder before starting the r10k checkout, but things are way slower if I have to check out all module repos in every run.

Did anyone see this warning before?

**Comment by haus** [comment]
That code comes from RK-265 (https://github.com/puppetlabs/r10k/pull/640). Do you think you could turn the loglevel up all the way to debug2 and give us the output?

**Comment by TheKangaroo** [comment]
Hey @haus , sorry for the delay. Here we go with loglevel debug2:
```
[2017-10-05 09:37:57 - DEBUG2] No config file explicitly given and no default config file could be found, default settings will be used.
[2017-10-05 09:37:57 - DEBUG1] Testing to see if feature pe_license is available.
[2017-10-05 09:37:57 - DEBUG2] Attempting to load library 'pe_license' for feature pe_license
[2017-10-05 09:37:57 - DEBUG2] Error while loading library pe_license for feature pe_license: cannot load such file -- pe_license
[2017-10-05 09:37:57 - DEBUG1] Feature pe_license is not available.
[2017-10-05 09:37:57 - DEBUG2] pe_license feature is not available, PE only Puppet modules will not be downloadable.
[2017-10-05 09:37:57 - DEBUG1] Testing to see if feature shellgit is available.
[2017-10-05 09:37:57 - DEBUG2] Evaluating proc #<Proc:0x00000002de04c0@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> to test for feature shellgit
[2017-10-05 09:37:57 - DEBUG2] Proc #<Proc:0x00000002de04c0@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> for feature shellgit returned "/bin/git"
[2017-10-05 09:37:57 - DEBUG1] Feature shellgit is available.
[2017-10-05 09:37:57 - DEBUG1] Testing to see if feature shellgit is available.
[2017-10-05 09:37:57 - DEBUG2] Evaluating proc #<Proc:0x00000002de04c0@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> to test for feature shellgit
[2017-10-05 09:37:57 - DEBUG2] Proc #<Proc:0x00000002de04c0@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> for feature shellgit returned "/bin/git"
[2017-10-05 09:37:57 - DEBUG1] Feature shellgit is available.
[2017-10-05 09:37:57 - DEBUG1] Setting Git provider to R10K::Git::ShellGit
[2017-10-05 09:37:57 - DEBUG1] Testing to see if feature shellgit is available.
[2017-10-05 09:37:57 - DEBUG2] Evaluating proc #<Proc:0x00000002de04c0@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> to test for feature shellgit
[2017-10-05 09:37:57 - DEBUG2] Proc #<Proc:0x00000002de04c0@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> for feature shellgit returned "/bin/git"
[2017-10-05 09:37:57 - DEBUG1] Feature shellgit is available.
[2017-10-05 09:37:57 - DEBUG1] Setting Git provider to default provider shellgit
[2017-10-05 09:37:57 - INFO] Updating module /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "rev-parse", "master^{commit}"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git rev-parse master^{commit}
Stdout:
d5436aec54b51162ea4e6ba7a54b0275e0f252ad
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "for-each-ref", "refs/heads", "--format", "%(refname)"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git for-each-ref refs/heads --format %(refname)
Stdout:
refs/heads/master
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "config", "--local", "--get-regexp", "^remote\\..*\\.url$"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git config --local --get-regexp ^remote\..*\.url$
Stdout:
remote.origin.url git@gitlab.private.com:puppet/a_module.git
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "fetch", "origin", "--prune"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git fetch origin --prune
Stderr:
From gitlab.private.com:puppet/a_module
   d5436ae..ced11dc  master     -> master
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "rev-parse", "master^{commit}"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git rev-parse master^{commit}
Stdout:
ced11dca83edc04dbaaf3ac1d8ca27842f3d4160
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "rev-parse", "master^{commit}"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git rev-parse master^{commit}
Stdout:
ced11dca83edc04dbaaf3ac1d8ca27842f3d4160
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "for-each-ref", "refs/heads", "--format", "%(refname)"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git for-each-ref refs/heads --format %(refname)
Stdout:
refs/heads/master
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "rev-parse", "master^{commit}"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git rev-parse master^{commit}
Stdout:
ced11dca83edc04dbaaf3ac1d8ca27842f3d4160
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git", "--work-tree", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module", "config", "--get", "remote.origin.url"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git --work-tree /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module config --get remote.origin.url
Stdout:
git@gitlab.private.com:puppet/a_module.git
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git", "--work-tree", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module", "diff-index", "--quiet", "HEAD", "--"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git --work-tree /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module diff-index --quiet HEAD --
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git", "rev-parse", "HEAD^{commit}"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git rev-parse HEAD^{commit}
Stdout:
d5436aec54b51162ea4e6ba7a54b0275e0f252ad
Exit code: 0
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "rev-parse", "master^{commit}"]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git rev-parse master^{commit}
Stdout:
ced11dca83edc04dbaaf3ac1d8ca27842f3d4160
Exit code: 0
[2017-10-05 09:37:57 - DEBUG] Updating /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module to master
[2017-10-05 09:37:57 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git", "--work-tree", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module", "checkout", "ced11dca83edc04dbaaf3ac1d8ca27842f3d4160", ""]
[2017-10-05 09:37:57 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git --work-tree /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module checkout ced11dca83edc04dbaaf3ac1d8ca27842f3d4160
Exit code: 0
[2017-10-05 09:37:57 - DEBUG1] No unmanaged contents in /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules, nothing to purge
## jenkins@infra-puppetdeploy-01:~/workspace/puppet_code_deploy/lxos_r10k_testing (till_testing) $ r10k puppetfile install -v
INFO	 -> Updating module /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module
WARN	 -> Skipping /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module due to local modifications
## jenkins@infra-puppetdeploy-01:~/workspace/puppet_code_deploy/lxos_r10k_testing (till_testing) $ r10k puppetfile install -v debug2
[2017-10-05 09:38:24 - DEBUG2] No config file explicitly given and no default config file could be found, default settings will be used.
[2017-10-05 09:38:24 - DEBUG1] Testing to see if feature pe_license is available.
[2017-10-05 09:38:24 - DEBUG2] Attempting to load library 'pe_license' for feature pe_license
[2017-10-05 09:38:24 - DEBUG2] Error while loading library pe_license for feature pe_license: cannot load such file -- pe_license
[2017-10-05 09:38:24 - DEBUG1] Feature pe_license is not available.
[2017-10-05 09:38:24 - DEBUG2] pe_license feature is not available, PE only Puppet modules will not be downloadable.
[2017-10-05 09:38:24 - DEBUG1] Testing to see if feature shellgit is available.
[2017-10-05 09:38:24 - DEBUG2] Evaluating proc #<Proc:0x00000001024620@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> to test for feature shellgit
[2017-10-05 09:38:24 - DEBUG2] Proc #<Proc:0x00000001024620@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> for feature shellgit returned "/bin/git"
[2017-10-05 09:38:24 - DEBUG1] Feature shellgit is available.
[2017-10-05 09:38:24 - DEBUG1] Testing to see if feature shellgit is available.
[2017-10-05 09:38:24 - DEBUG2] Evaluating proc #<Proc:0x00000001024620@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> to test for feature shellgit
[2017-10-05 09:38:24 - DEBUG2] Proc #<Proc:0x00000001024620@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> for feature shellgit returned "/bin/git"
[2017-10-05 09:38:24 - DEBUG1] Feature shellgit is available.
[2017-10-05 09:38:24 - DEBUG1] Setting Git provider to R10K::Git::ShellGit
[2017-10-05 09:38:24 - DEBUG1] Testing to see if feature shellgit is available.
[2017-10-05 09:38:24 - DEBUG2] Evaluating proc #<Proc:0x00000001024620@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> to test for feature shellgit
[2017-10-05 09:38:24 - DEBUG2] Proc #<Proc:0x00000001024620@/var/lib/jenkins/.gem/ruby/gems/r10k-2.5.5/lib/r10k/features.rb:16> for feature shellgit returned "/bin/git"
[2017-10-05 09:38:24 - DEBUG1] Feature shellgit is available.
[2017-10-05 09:38:24 - DEBUG1] Setting Git provider to default provider shellgit
[2017-10-05 09:38:24 - INFO] Updating module /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module
[2017-10-05 09:38:24 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "rev-parse", "master^{commit}"]
[2017-10-05 09:38:24 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git rev-parse master^{commit}
Stdout:
ced11dca83edc04dbaaf3ac1d8ca27842f3d4160
Exit code: 0
[2017-10-05 09:38:24 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "for-each-ref", "refs/heads", "--format", "%(refname)"]
[2017-10-05 09:38:24 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git for-each-ref refs/heads --format %(refname)
Stdout:
refs/heads/master
Exit code: 0
[2017-10-05 09:38:24 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "config", "--local", "--get-regexp", "^remote\\..*\\.url$"]
[2017-10-05 09:38:24 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git config --local --get-regexp ^remote\..*\.url$
Stdout:
remote.origin.url git@gitlab.private.com:puppet/a_module.git
Exit code: 0
[2017-10-05 09:38:24 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "fetch", "origin", "--prune"]
[2017-10-05 09:38:24 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git fetch origin --prune
Exit code: 0
[2017-10-05 09:38:24 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "rev-parse", "master^{commit}"]
[2017-10-05 09:38:24 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git rev-parse master^{commit}
Stdout:
ced11dca83edc04dbaaf3ac1d8ca27842f3d4160
Exit code: 0
[2017-10-05 09:38:24 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "rev-parse", "master^{commit}"]
[2017-10-05 09:38:24 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git rev-parse master^{commit}
Stdout:
ced11dca83edc04dbaaf3ac1d8ca27842f3d4160
Exit code: 0
[2017-10-05 09:38:24 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "for-each-ref", "refs/heads", "--format", "%(refname)"]
[2017-10-05 09:38:24 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git for-each-ref refs/heads --format %(refname)
Stdout:
refs/heads/master
Exit code: 0
[2017-10-05 09:38:24 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git", "rev-parse", "master^{commit}"]
[2017-10-05 09:38:24 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/.r10k/git/git@gitlab.private.com-puppet-a_module.git rev-parse master^{commit}
Stdout:
ced11dca83edc04dbaaf3ac1d8ca27842f3d4160
Exit code: 0
[2017-10-05 09:38:24 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git", "--work-tree", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module", "config", "--get", "remote.origin.url"]
[2017-10-05 09:38:24 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git --work-tree /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module config --get remote.origin.url
Stdout:
git@gitlab.private.com:puppet/a_module.git
Exit code: 0
[2017-10-05 09:38:24 - DEBUG2] Starting process: ["git", "--git-dir", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git", "--work-tree", "/var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module", "diff-index", "--quiet", "HEAD", "--"]
[2017-10-05 09:38:24 - DEBUG2] Finished process:
Command: git --git-dir /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module/.git --work-tree /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module diff-index --quiet HEAD --
Exit code: 1
[2017-10-05 09:38:24 - WARN] Skipping /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules/a_module due to local modifications
[2017-10-05 09:38:24 - DEBUG1] No unmanaged contents in /var/lib/jenkins/workspace/puppet_code_deploy/lxos_r10k_testing/modules, nothing to purge
```

I did the following:
- create the Puppetfile with content
```
mod 'a_module',
  :git => 'git@gitlab.private.com:puppet/a_module.git',
  :ref => 'master'
```
- make sure the modules dir is completely empty
- deploy with "r10k puppetfile install -v"
- updated the module 
- don't change the moudle under modules/a_module
- deploy with r10k puppetfile install -v debug2


The refs matches the two in the log output:
old ref d5436aec54b51162ea4e6ba7a54b0275e0f252ad
new ref ced11dca83edc04dbaaf3ac1d8ca27842f3d4160

the module is still on the old commit:
```
# cd modules/a_module
# git log --de
commit d5436aec54b51162ea4e6ba7a54b0275e0f252ad (HEAD, tag: 1.0.0, origin/master, origin/HEAD, cache/master, master)
...
```

Looks very strange to me.

**Comment by haus** [comment]
Thanks @TheKangaroo !

**Comment by TheKangaroo** [comment]
Hey @haus,
any news on that? This really slows down our puppet code deployment process.

**Comment by andersonmills** [comment]
@TheKangaroo - we've got a PR up for an --update_force flag for the puppetfile install action and subcommand. #772

**Comment by andersonmills** [comment]
#772 is now merged and will be going into the 2.5.x series.

**Comment by TheKangaroo** [comment]
@andersonmills nice to hear that.
Anyway, I'm curious how this could even happen, since I didn't change a thing in the checked out repo. In this case, I shouldn't even need the --update_fore flag.
The --update_force will work for me but might not resolve the problem for other people.

**Comment by andersonmills** [comment]
@TheKangaroo Are you asking "how could it ever happen that a behavior change could be added to a .Y release?" The change in behavior shouldn't have happened in a .Y, but it did. The idea from the beginning is that ```r10k puppetfile install``` was more for people in dev mode, so don't overwrite local changes as default made sense from that mindset.

This issue should be closable with the upcoming release of 2.5.6.

**Comment by andersonmills** [comment]
Be aware that this is still in development and I might get asked to change the flag name before the release.

**Comment by TheKangaroo** [comment]
@andersonmills I'm sorry for my confusing answer. I wasn't talking about the changes in r10k, but rather the (allegedly) changes in my checked out git repo.
What I wanted to say is that git in the Jenkins job does not act like I expected it to do.

Since I don't change a thing in the checked out repo I am confused how git could think there are local changes.
As far as I can see is this an issue outside the scope of r10k. Nethertheless I think the --update_force flag will work around this for me.

**Comment by jarro2783** [comment]
I think this is fixed in 49b8b068b8b802cc8ca9daeca84b795ba6c00455, which is dated 2016-12-12, but still hasn't made it into a released version.

When you update a git repo it ends up doing `git checkout ref ""`, which actually checks out the commit with staged changes. On the next run it thinks there are local changes.

Hopefully this commit ends up in version 3 since it looks like that has been merged very recently.

**Comment by Magisus** [comment]
I think this has been released now. Closing.
